src/samples/http/server-sse.c

```diff
@@ -0,0 +1,295 @@
+static int port = 8080;
+
+static bool
+on_stdin(void *data, int fd, uint32_t flags)
+{
+    uint16_t i;
+    struct sol_http_response_progressive *sse;
+    static const struct sol_str_slice begin = {
+        .len = sizeof("data: ") - 1,
+        .data = (const char *)"data: "
+    };
+    static const struct sol_str_slice end = {
+        .len = sizeof("\n\n") - 1,
+        .data = (const char *)"\n\n"
+    };
+
+    if (flags & (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_HUP)) {
+        fprintf(stderr, "ERROR: Something wrong happened with file descriptor: %d\n", fd);
+        goto err;
+    }
+
+    if (flags & SOL_FD_FLAGS_IN) {
+        int err;
+
+        value.used = 0;
+        /* this will loop trying to read as much data as possible to buffer. */
+        err = sol_util_load_file_fd_buffer(fd, &value);
+        if (err < 0) {
+            fprintf(stderr, "ERROR: failed to read from stdin: %s\n",
+                sol_util_strerrora(-err));
+            goto err;
+        }
+
+        if (value.used == 0) {
+            /* no data usually means ^D on the terminal, quit the application */
+            printf("no data on stdin, quitting.\n");
+            sol_quit();
+        } else {
+            err = sol_buffer_insert_slice(&value, 0, begin);
+            if (err < 0) {
+                fprintf(stderr, "ERROR: failed to append the data prefix: %s\n",
+                    sol_util_strerrora(-err));
+                goto err;
+            }
+
+            err = sol_buffer_append_slice(&value, end);
+            if (err < 0) {
+                fprintf(stderr, "ERROR: Could not append data to buffer: %s\n",
+                    sol_util_strerrora(-err));
+                goto err;
+            }
+        }
+    }
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&responses, sse, i)
+        sol_http_response_progressive_feed(sse, sol_buffer_get_slice(&value));
+
+    return true;
+
+err:
+    sol_quit_with_code(EXIT_FAILURE);
+    stdin_watch = NULL;
+    return false;
+}
+
+static void
+delete_cb(void *data, const struct sol_http_response_progressive *sse)
+{
+    sol_ptr_vector_remove(&responses, sse);
+}
+
+static int
+request_events_cb(void *data, struct sol_http_request *request)
+{
+    int ret;
+    struct sol_http_response_progressive *sse;
+    struct sol_http_response response = {
+        SOL_SET_API_VERSION(.api_version = SOL_HTTP_RESPONSE_API_VERSION, )
+        .param = SOL_HTTP_REQUEST_PARAMS_INIT,
+        .response_code = SOL_HTTP_STATUS_OK,
+        .content = SOL_BUFFER_INIT_EMPTY
+    };
+
+    ret = sol_http_response_set_progressive_headers(&response);
+    if (ret < 0)
+        return ret;
+
+    sse = sol_http_server_send_response_progressive(request, &response, delete_cb, NULL);
+    sol_http_params_clear(&response.param);
+    if (!sse)
+        return -1;
+
+    ret = sol_ptr_vector_append(&responses, sse);
+    if (ret < 0) {
+        sol_http_response_progressive_del(sse, false);
```

Comment:
I'm not sure this part will not cause a double free, try to force this error and confirm with valgrind.

`sol_http_response_progressive_del()` will just mark, but will keep the `sse` alive, but then this function returns an error and `request` will be deleted. Please confirm with a forced error if that works as expected.